pkg/ccl/sqlproxyccl/acl/watcher.go

```diff
@@ -173,14 +160,32 @@ func (l *listener) Less(than btree.Item) bool {
 	return l.id < than.(*listener).id
 }
 
-func checkConnection(connection ConnectionTags, list *Denylist) error {
-	ip := DenyEntity{Item: connection.IP, Type: IPAddrType}
-	if err := list.Denied(ip); err != nil {
-		return errors.Wrapf(err, "connection ip '%v' denied", connection.IP)
-	}
-	cluster := DenyEntity{Item: connection.Cluster, Type: ClusterType}
-	if err := list.Denied(cluster); err != nil {
-		return errors.Wrapf(err, "connection cluster '%v' denied", connection.Cluster)
+func checkListeners(
+	listeners *btree.BTree, timesource timeutil.TimeSource, controllers ...AccessController,
```

Comment:
Ah my thought here was that by passing them in individually rather than the slice, I would avoid races. But you're right, I would need to make a copy of the slice anyways so I will just do that instead.